source/common/config/grpc_mux_impl.cc

```diff
@@ -355,10 +386,24 @@ void GrpcMuxImpl::processDiscoveryResources(const std::vector<DecodedResourcePtr
       continue;
     }
     std::vector<DecodedResourceRef> found_resources;
+    absl::flat_hash_set<std::string> added_resources;
     for (const auto& watched_resource_name : watch->resources_) {
       auto it = resource_ref_map.find(watched_resource_name);
       if (it != resource_ref_map.end()) {
         found_resources.emplace_back(it->second);
+        added_resources.insert(watched_resource_name);
```

Comment:
I'm not sure I fully understand why `added_resources` is needed.
It seems that both wildcard and non-wildcard watchers will need to have the resource as part of the `found_resources`, no?

nit: in the TODO I think it would be better to have a dedicated map of wildcard watches, so the search can be inverted (go over the resources, and find the wildcard-watches that is relevant). WDYT?

Reply:
The reason for added_resources is because if we have a watch on a wildcard, we loop through the resource_ref_map, so we could end up adding the same entry from the resource_ref_map more than once.  E.g. if we are watching `xdstp://test/foo` and `xdstp://test/*`, and the resource_ref_map contains `xdstp://test/foo`.  However, instead of having the set keep track of whether we've already inserted it, I just did an erase from the resource_ref_map once we've inserted into found_resources.

RE the TODO, sorry I think I wasn't clear, what I meant is basically what you're suggesting.  I tried clearing up the language.b

Reply:
I guess my comment wasn't clear. In the case of watching `xdstp://test/foo` and `xdstp://test/*` why aren't both watchers notified with the resource? I guess it could also raise another question: in the current implementation why does singleton-watcher take precedence over collection-watcher?

Reply:
Discussed offline, removed the deduping check as we should not have singleton and collection resources in the same watch; and I commented the assumption.